scred/backfillna.py

```diff
@@ -26,116 +33,147 @@
+        return False  # Handles blank result from key
+
 
 cond.setParseAction(check_condition)
 
 
 def fullparse(expression):
     """
-    Takes in an expression and parses it using the full branching logic. This
-    attempts to split the string into tokens, put the tokens back together
-    in a string with responses instead of field names, and evaluate that string.
+    Takes in an expression and parses it using the full branching logic.
+    This attempts to split the string into tokens, put the tokens back
+    together in a string with responses instead of field names, and
+    evaluate that string.
     """
     global logic
-    try: 
+    try:
         parsed_expr = logic.parseString(expression)
         parsed_expr = " ".join([str(x) for x in parsed_expr])
         return bool(eval(parsed_expr))
     except KeyError:
         # If condition exists but can't be parsed, assume not met
-        print("WARNING! KeyError during Parser.fullparse(), may falsely code as N/A.")
-        return False 
+        print(
+            "WARNING! KeyError during Parser.fullparse(),"
+            "may falsely code as N/A."
+        )
+        return False
     except pp.ParseException:
         # This means there was no logic, so we accept it as met
         return True
 
+
 # ===================================================
 
+
 class Parser:
     def __init__(self, data):
-        global key, operation, value, cond, joint, cond_chain, cond_chain_with_parentheses, logic
+        global key, operation, value, cond, joint
+        global cond_chain, cond_chain_with_parentheses, logic
         self.data = data
-        self.data["LOGIC_MET"] = "" # Add empty column for logic result
+        self.data["LOGIC_MET"] = ""  # Add empty column for logic result
         # Complete parser setup by pointing action at this instance
         key.setParseAction(self.use_key)
 
-
     def use_key(self, list_with_key):
         """
         Access a key's value in the instance data.
         """
         k = list_with_key[0]
         try:
-            return self.val_from_key(k) 
+            return self.val_from_key(k)
         except KeyError:
-            warnings.warn(f"WARNING: Caught KeyError in Parser.use_key() for {k}")
+            warnings.warn(
+                f"WARNING: Caught KeyError in Parser.use_key() for {k}"
+            )
             return False
 
-
     def val_from_key(self, key, data=None):
         """
-        Helper function that takes a key, looks that key up in the instance record, then returns its value.
-        Not sure how this works with use_key and giving a list and all that, but it does. Should revisit.
+        Helper function that takes a key, looks that key up in the
+        instance record, then returns its value. Not sure how this
+        works with use_key and giving a list and all that, but it does.
+        Should revisit.
         """
-        if data == None:
+        if data is None:
```

Comment:
This is a funky one--there's a bug, or maybe quirk is a better word, in `pandas` that makes comparisons to `None` work differently than in base python.  My understanding is as follows: `pandas` wants to return a `pd.Series` when getting values, even if there's only one value. Since `data` is a `pd.DataFrame`, there are situations where it acts like a series containing `np.nan` where we'd expect to see None; also, `np.nan == np.nan` is False. This can give surprising results with element-wise comparisons, even if there's only one element. `Parser.use_key` and `check_condition` both pull the 0th element of a one-value container.

I'm trying to figure out if that situation actually applies here. If not, we should change this to your suggestion. If it does, we should still do something to make it more explicit; this is a confusing piece of code! The reference I have saved is https://stackoverflow.com/questions/58190953/why-does-pandas-isnull-work-but-none-not-work

What do you think? I'll poke around at mixing pandas with pyparsing; I think there was something with their default return values not liking each other, just need to get hands-on with this code again to remember the details. I also did this back in 0.22.x, so it might be different now that we're using 0.25.3.

Reply:
Interesting. I'll look into this more. The principle I was following (and that `pylama` warns about is actually quoted in the SO post you included ("Comparisons to singletons like None should always be done with `is` or `is not`, never the equality operators"), but it sounds like you're referring to a quirk that isn't quite the same as that SO post.